### ISSUE TITLE
process manager, wait for another process manager

### ISSUE THREAD

**Issue by beckend** [redux-observable/redux-observable#63]
I haven't figured out a way to wait for another process manager inside a process manager,
store dispatch simply returns the action object which is of no use.

```javascript

const fetchWhatEverManager = (action$, store) =>
  action$.ofType(FETCH_USER)
    .switchMap(action =>
      const maybeData = store.getState().getIn([ 'myReducer', 'myData' ]);
      if (!store.getState().getIn([ 'myReducer', 'myData' ])) {
        // Data has not been fetched to continue, need to fetch it right here
        // What to do here??? Dispatch does not wait for action
        const onlyActionObj = store.dispatch({ type: 'FETCH_MY_DATA_AND_POPULATE_REDUCER' });
      } else {
        // Did not need to fetch, do something with available data and return payload
        const payload = maybeData................
        return Observable.of({ type: 'MY_ACTION', payload });
      }
    );
```

**Comment by matthewwithanm** [comment]
Process managers are just observable-returning functions so you could just call the one that handles `FETCH_MY_DATA_AND_POPULATE_REDUCER` directly.

**Comment by jayphelps** [comment]
I can definitely provide a solution for you, but I need to know more specifically what you're trying to accomplish. What is the _real_ action you wish to dispatch for each branch and what do they do? The reason I need to know is that this is almost certainly just structured improperly, but the way to structure it requires knowing what result you really want, not how you _currently_ think it should be done.

**Comment by beckend** [comment]
Ok let me try to explain.
So it's some kind of web hosting application, user can view their sites and pages.

1. User enter application, a clean state. There are some menus. User however has access token from cookies.
2. User decides to click on view pages.

But pages needs sites(a separate endpoint) to be fetches first.

I know I could do something in the manager for fetching `sites`:
```javascript
if (currentPath === '/pages') {
  doFetchPages();
}
```
But I would rather have the logic inside `doFetchPages`, since `sites` are not always relevant.

Basically, there are routes that depends on higher data to be available, but I really don't want to fetch it at application start when it's not really needed

I'll try to do what @matthewwithanm suggested.

**Comment by beckend** [comment]
So I gave it a try, it seems successful, just a lot of boilerplate.
```javascript
const fetchWhatEverManager = (action$, store) =>
  action$.ofType(FETCH_MYSUBDATA)
    .switchMap(action => {
      const maybeData = store.getState().getIn([ 'myReducer', 'myData' ]);
      if (!maybeData) {
        // Fetch dependent data
        // Bluebird
        let mySub;
        const myPromise = new Promise((resolve, reject) => {
          mySub = myOtherFetchManager(action$)
            .subscribe(theAction => {
              if (theAction.type === 'FETCH_TOP_DATA_SUCCESS') {
                resolve(theAction.payload);
              }
              if (theAction.type === 'FETCH_TOP_DATA_FAILURE') {
                reject(new Error(theAction.payload));
              }
            });
        }).finally(() => {
          mySub.unsubscribe();
        });


        // Consume
        return Observable.fromPromise(myPromise)
          .map(() => { ... });

      } else {
        // Did not need to fetch, do something with available data and return payload
        const payload = maybeData................
        return Observable.of({ type: 'FETCH_TOP_DATA', payload });
      }
    });

```

**Comment by matthewwithanm** [comment]
Making subscriptions like that generally isn't something you need to do with Rx. Since the other process manager returns an observable of actions you should be able to just switchMap to it (or something you drive from it).

**Comment by jayphelps** [comment]
@beckend hmmm I'm still just as lost because I don't get why in your example code you have a condition based on `maybeData` but what data is this from the store? `myReducer` and `myData` aren't descriptive at all and what is "do something with available data and return payload"? Why does it appear that at the end of the two branches you're not dispatching the same action? Basically--why do you need to do one thing if the data exists in the store but do something completely different if not?

The description you gave of the use case seems pretty trivial--when you need the "sites" data, you fetch it. When you don't, you don't. You wouldn't need path-based checks if you're using something like react-router. You would `dispatch(fetchSites())` or whatever you need inside the applicable lifecycle hooks of the route component you navigated to that needs that data...e.g. `componentWillMount` (and possibly in `componentWillUpdate`, depending).

**Comment by beckend** [comment]
@matthewwithanm 
Thats what I did first, tried to chain and use it, but unable to just yet figure out how to get it working, so at first I did:
```javascript
myOtherFetchManager(action$)
  .switchMap(() => { ... })
```
But the code was a dead end, I'm still working on it, that would be the most optimal solution without needing to subscribe/cleanup.

@jayphelps
Maybe my example sucked, which I'm sorry for.
My current application has a lot of components, actually I'm not even using the routes.

So a component `A` does have on `componentWillMount` `this.props.fetchDATA_A();`.

And another component `B` does have on `componentWillMount` `this.props.fetchDATA_B();`

The payload for the second component does depends on `fetchDATA_A`, to calculate the payload needed, which I can get from the store reducer if it's fetched.

Both of the components are not used together, they are their completely isolated and works independently, so that's why I opted for this solution.

A possible solution was to use on component `B`:
```javascript
componentWillMount() {
  if (!this.context.store.getIn([ 'myReducer', 'data' ])) {
    // Chain and await fetch fetchDATA_A, then fetch fetchDATA_B
  } else {
    this.props.fetchDATA_B();
  }
}
```
But since using process manager, this was not possible I guess?
And is only possible to use with "thunkservables"?
```javascript
dispatch(() => Rx.Observable.of({ type: 'ASYNC_ACTION_A' })
  .delay(1000))
  .map(() => { type: 'ASYNC_ACTION_B' });
```

**Comment by jayphelps** [comment]
@beckend My issue with the examples continues to be that they are pseudo code, assuming you're not really naming your variables `myReducer`, `data`, `fetchDATA_A` etc. The best way to do this really does depend on _what the data actually is and how they actually relate to each other_.

That said, a very generic example (that may not be the best, depending on the data) looks something like this: (totally untested, but should give you the jist)

```js

const fetchDataA = () => ({ type: FETCH_DATA_A });
const fetchDataAFulfilled = payload => ({ type: FETCH_DATA_A_FULFILLED, payload });
const fetchDataB = () => ({ type: FETCH_DATA_B });
const fetchDataBFulfilled = payload => ({ type: FETCH_DATA_B_FULFILLED, payload });

const fetchDataAFromAPI = () =>
  ajax('/api/path/for/data/a');

const fetchDataAManager = (action$ store) =>
  action$.ofType(FETCH_DATA_A)
    .switchMap(action =>
      fetchDataAFromAPI()
        .map(fetchDataAFulfilled)
    );

const fetchDataBManager = (action$ store) =>
  action$.ofType(FETCH_DATA_B)
    .switchMap(action => {
      const maybeData = store.getState().getIn(['myReducer', 'myData']);
      if (!maybeData) {
        return fetchDataAFromAPI()
          .map(payload => ({
            // normalize the result to be the same as the data from the store
            myReducer: payload.where_ever_it_is,
            myData: payload.where_ever_it_is
          }));
      } else {
        return Observable.of(maybeData);
      }
    .switchMap(data =>
      ajax(`/api/path/for/data/b/${data.myData}`)
        .map(fetchDataBFulfilled)
    );
```

Again, I'd need to know the actual data you're dealing with to provide the best example. For many cases, I personally think this logic does indeed belong in the components themselves, instead of in the process manager. That is, the component should get the values from the store, and if they don't exist yet, fetch them before proceeding onto the next item to fetch that depends on it. I don't have any good argument _why_ I believe that, just a feeling. Definitely feel free to continue keeping it in the process manager, if you find this is more maintainable.

**Comment by beckend** [comment]
That will work, thanks, I'm still getting around process managers.
